File: ios/Classes/CharacterPlatformView.swift
import Flutter
import UIKit
import SPAvatarKit

class CharacterPlatformViewFactory: NSObject, FlutterPlatformViewFactory {
    private var messenger: FlutterBinaryMessenger
    
    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }
    
    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        var sessionToken = ""
        var customViewId: Int64 = viewId
        var channelName = ""
        
        if let args = args as? [String: Any] {
            if let viewIdParam = args["viewId"] as? Int64 {
                customViewId = viewIdParam
            }
            if let token = args["sessionToken"] as? String {
                sessionToken = token
            }
            if let channel = args["channelName"] as? String {
                channelName = channel
            }
        }
        
        return CharacterPlatformView(
            frame: frame,
            sessionToken: sessionToken,
            viewIdentifier: customViewId,
            channelName: channelName,
            messenger: messenger
        )
    }
    
    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

class CharacterPlatformView: NSObject, FlutterPlatformView, CharacterViewControllerDelegate {
    private var _characterView: UIView
    private var _characterVC: CharacterViewController
    private var channel: FlutterMethodChannel

    init(frame: CGRect, sessionToken: String, viewIdentifier viewId: Int64, channelName: String, messenger: FlutterBinaryMessenger) {
        _characterVC = CharacterViewController()
        _characterView = _characterVC.view
        channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
        super.init()
        _characterVC.delegate = self
        _characterVC.view.frame = frame
        
        AvatarSDK.shared.setupIfNeeded(sessionToken: sessionToken)
        
        channel.setMethodCallHandler { [weak self] call, result in
            self?.handleFlutterCall(call, result: result)
        }
    }
    
    func view() -> UIView {
        return _characterView
    }
    
    func handleFlutterCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let method = PlatformMethod(rawValue: call.method) else {
            print("❌ CharacterPlatformView: 未知方法: \(call.method)")
            return result(FlutterMethodNotImplemented) 
        }
        switch method {
        case .loadCharacterData:
            if let dict = call.arguments as? [String: Any], let characterId = dict["characterId"] as? String {
                let isBackgroundOpaque = dict["isBackgroundOpaque"] as? Bool ?? true
                if let backgroundImage = dict["backgroundImage"] as? FlutterStandardTypedData {
                  Task {
                      let image = UIImage(data: backgroundImage.data)
                      await _characterVC.loadCharacter(characterId, backgroundImage: image, isBackgroundOpaque: isBackgroundOpaque)
                  }
                } else {
                  Task {
                    await _characterVC.loadCharacter(characterId, backgroundImage: nil, isBackgroundOpaque: isBackgroundOpaque)
                  }
                }
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_PATH", message: "Expected Map path", details: nil))
            }
        case .preloadCharacterData:
            if let characterId = call.arguments as? String {
                Task {
                    let success = await _characterVC.preloadCharacter(characterId)
                    result(success)
                }
                result(nil)
            }  else {
                result(FlutterError(code: "INVALID_PATH", message: "Expected String path", details: nil))
            }
        case .deleteCharacterAssets:
            if let characterId = call.arguments as? String {
                Task {
                     await _characterVC.deleteCharacterAssets(characterId)
                }
                result(nil)
            }  else {
                result(FlutterError(code: "INVALID_PATH", message: "Expected String path", details: nil))
            }
        case .deleteAllCharacterAssets:
            Task {
                await _characterVC.deleteAllCharacterAssets()
            }
            result(nil)
        case .start:
            _characterVC.start()
            result(nil)
        case .close:
            if let dict = call.arguments as? [String: Any], let shouldCleanup = dict["shouldCleanup"] as? Bool {
                _characterVC.close(shouldCleanup: shouldCleanup)
            } else {
                _characterVC.close()
            }
            result(nil)
        case .interrupt:
            _characterVC.interrupt()
            result(nil)
        case .sendAudioData:
            guard AvatarSDK.shared.setUpState == .successed else {
                return result(FlutterError(code: "SDK_NOT_SETUP", message: "SDK setup not completed", details: nil))
            }
            if let dict = call.arguments as? [String: Any], let data = dict["audioData"] as? FlutterStandardTypedData, let end = dict["end"] as? Bool {
                let audioData = data.data
                let ret = _characterVC.sendAudioData(audioData, end: end)
                result(ret)
            } else {
                result(FlutterError(code: "INVALID_DATA", message: "Expected Uint8List", details: nil))
            }
        case .setVolume:
            if let volume = call.arguments as? Double {
                _characterVC.setVolume(Float(volume))
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_DATA", message: "Expected Double", details: nil))
            }
        }
    }
    
    func didReceiveEvent(event: PlatformEvent, params: Any) {
        // 当事件发生时，通过MethodChannel通知Flutter
        print("🔧 CharacterPlatformView: 发送事件到Flutter，事件: \(event.rawValue), 参数: \(params)")
        channel.invokeMethod(event.rawValue, arguments: params)
    }
    
    // MARK: - CharacterViewControllerDelegate
    
    func characterViewController(_ characterViewController: CharacterViewController, didReceivedEvent event: PlatformEvent, params: Any) {
        didReceiveEvent(event: event, params: params)
    }
}

class AvatarSDK: SPAvatarSDKDelegate {
    static let shared = AvatarSDK()

    private init() {}

    private var onceToken = false

    func setupIfNeeded(sessionToken: String) {
        guard !onceToken else { return }
        onceToken = true
        SPAvatarSDK.shared.setup(sessionToken: sessionToken, configuration: SPAvatarSDK.Configuration(), delegate: self)
    }

    private(set) var setUpState: PlatformSetUpState = .successed

    func avatarSDKDidStarted() {
        setUpState = .successed
    }
    
    func avatarSDKFailedToStart() {
        setUpState = .notSetUp
    }
}

File: android/src/main/kotlin/ai/spatialwalk/sp_character_plugin/CharacterPlatformView.kt
package ai.spatialwalk.sp_character_plugin

import android.content.Context
import android.graphics.Color
import android.util.Log
import android.view.View
import android.widget.FrameLayout
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import kotlinx.coroutines.*

/**
 * CharacterPlatformView - manages the native Android view for character display
 */
class CharacterPlatformView(
    private val context: Context,
    private val messenger: BinaryMessenger,
    private val viewId: Int,
    sessionToken: String,
    channelName: String
) : PlatformView, CharacterViewControllerDelegate {
    
    companion object {
        private const val TAG = "AvatarKitCharacterPlatformView"
    }

    private val containerView: FrameLayout = FrameLayout(context)
    private val methodChannel: MethodChannel = MethodChannel(messenger, channelName)
    // Lifecycle-aware coroutine scope that runs on background thread
    // SupervisorJob ensures failure in one coroutine doesn't cancel others
    private val coroutineScope = CoroutineScope(Dispatchers.Default + SupervisorJob())
    private val characterViewController: CharacterViewController

    init {
        Log.d(TAG, "Initializing CharacterPlatformView with viewId: $viewId, channelName: $channelName")
        
        characterViewController = CharacterViewController(context, coroutineScope)
        characterViewController.delegate = this
        
        // Setup method channel handler
        Log.d(TAG, "Setting up MethodChannel handler for: $channelName")
        methodChannel.setMethodCallHandler { call, result ->
            handleFlutterCall(call, result)
        }
        
        // Initialize SDK if needed
        AvatarSDK.setupIfNeeded(context)
        AvatarSDK.setSessionToken(sessionToken)
        
        // Notify setup state
        onReceivedEvent(PlatformEvent.DID_UPDATED_CONNECTION_STATE, "disconnected")
    }

    override fun getView(): View {
        return containerView
    }

    override fun dispose() {
        Log.d(TAG, "Disposing CharacterPlatformView")
        characterViewController.cleanup()
        coroutineScope.cancel()
        methodChannel.setMethodCallHandler(null)
    }

    /**
     * Handle Flutter method calls
     */
    private fun handleFlutterCall(call: MethodCall, result: MethodChannel.Result) {
        Log.d(TAG, "Received Flutter call: ${call.method}")
        val method = PlatformMethod.fromValue(call.method)
        
        if (method == null) {
            Log.w(TAG, "Unknown method: ${call.method}")
            result.notImplemented()
            return
        }

        when (method) {
            PlatformMethod.LOAD_CHARACTER_DATA -> {
                handleLoadCharacter(call, result)
            }
            PlatformMethod.PRELOAD_CHARACTER_DATA -> {
                handlePreloadCharacter(call, result)
            }
            PlatformMethod.DELETE_CHARACTER_ASSETS -> {
                handleDeleteCharacterAssets(call, result)
            }
            PlatformMethod.DELETE_ALL_CHARACTER_ASSETS -> {
                handleDeleteAllCharacterAssets(result)
            }
            PlatformMethod.START -> {
                characterViewController.start()
                result.success(null)
            }
            PlatformMethod.CLOSE -> {
                val shouldCleanup = (call.arguments as? Map<*, *>)?.get("shouldCleanup") as? Boolean ?: false
                characterViewController.close(shouldCleanup)
                result.success(null)
            }
            PlatformMethod.INTERRUPT -> {
                characterViewController.interrupt()
                result.success(null)
            }
            PlatformMethod.SEND_AUDIO_DATA -> {
                handleSendAudioData(call, result)
            }
            PlatformMethod.SET_VOLUME -> {
                val volume = call.arguments as? Double
                if (volume != null) {
                    characterViewController.setVolume(volume.toFloat())
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "Volume must be a number", null)
                }
            }
        }
    }

    /**
     * Handle load character request
     */
    private fun handleLoadCharacter(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *>
        val characterId = args?.get("characterId") as? String
        val backgroundImage = args?.get("backgroundImage") as? ByteArray
        val isBackgroundOpaque = args?.get("isBackgroundOpaque") as? Boolean ?: true

        if (characterId == null) {
            result.error("INVALID_ARGUMENT", "characterId is required", null)
            return
        }

        
        // Add avatar view to container when available
        coroutineScope.launch(Dispatchers.IO) {
            // Load character
            characterViewController.loadCharacter(characterId, backgroundImage, isBackgroundOpaque)
            characterViewController.getView()?.let { avatarView ->
                withContext(Dispatchers.Main) {
                    containerView.removeAllViews()
                    containerView.addView(
                        avatarView,
                        FrameLayout.LayoutParams(
                            FrameLayout.LayoutParams.MATCH_PARENT,
                            FrameLayout.LayoutParams.MATCH_PARENT
                        )
                    )
                }
            }
            result.success(null)
        }
    }

    /**
     * Handle preload character request
     */
    private fun handlePreloadCharacter(call: MethodCall, result: MethodChannel.Result) {
        val characterId = call.arguments as? String
        
        if (characterId == null) {
            result.error("INVALID_ARGUMENT", "characterId is required", null)
            return
        }

        coroutineScope.launch {
            val success = characterViewController.preloadCharacter(characterId)
            result.success(success)
        }
    }

    /**
     * Handle delete character assets request
     */
    private fun handleDeleteCharacterAssets(call: MethodCall, result: MethodChannel.Result) {
        val characterId = call.arguments as? String
        
        if (characterId == null) {
            result.error("INVALID_ARGUMENT", "characterId is required", null)
            return
        }

        coroutineScope.launch {
            characterViewController.deleteCharacterAssets(characterId)
            result.success(null)
        }
    }

    /**
     * Handle delete all character assets request
     */
    private fun handleDeleteAllCharacterAssets(result: MethodChannel.Result) {
        coroutineScope.launch {
            characterViewController.deleteAllCharacterAssets()
            result.success(null)
        }
    }

    /**
     * Handle send audio data request
     */
    private fun handleSendAudioData(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *>
        val audioData = args?.get("audioData") as? ByteArray
        val end = args?.get("end") as? Boolean ?: false

        if (audioData == null) {
            result.error("INVALID_ARGUMENT", "audioData is required", null)
            return
        }

        val ret = characterViewController.sendAudioData(audioData, end)
        result.success(ret)
    }

    /**
     * CharacterViewControllerDelegate implementation
     */
    override fun onReceivedEvent(event: PlatformEvent, params: Any) {
        Log.d(TAG, "Received event: ${event.value}, params: $params")
        coroutineScope.launch(Dispatchers.Main) {
            methodChannel.invokeMethod(event.value, params)
        }
    }
}

/**
 * Singleton for SDK initialization
 */
object AvatarSDK {
    private const val TAG = "AvatarSDK"
    private var isInitialized = false
    private var environment: ai.spatialwalk.avatarkit.AvatarKit.Environment =
        ai.spatialwalk.avatarkit.AvatarKit.Environment.TEST

    @Synchronized
    fun setupIfNeeded(context: Context) {
        if (isInitialized) {
            return
        }

        try {
            Log.d(TAG, "Initializing AvatarKit SDK")
            ai.spatialwalk.avatarkit.AvatarKit.initialize(
                context = context,
                appId = "", // App ID should be set by the application
                configuration = ai.spatialwalk.avatarkit.AvatarKit.Configuration(
                    environment = environment
                )
            )
            isInitialized = true
            Log.d(TAG, "AvatarKit SDK initialized successfully")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize AvatarKit SDK: ${e.message}", e)
        }
    }

    fun setEnvironment(env: ai.spatialwalk.avatarkit.AvatarKit.Environment) {
        environment = env
    }

    fun setSessionToken(token: String) {
        ai.spatialwalk.avatarkit.AvatarKit.sessionToken = token
    }
}

